{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import requests\n",
    "from zipfile import ZipFile\n",
    "from ahocorasick import Automaton\n",
    "from tqdm.auto import tqdm"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Code to clean the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/jb/ntxm50wx2wv90pnq36yd8w940000gn/T/ipykernel_60618/804822752.py:4: DtypeWarning: Columns (7) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  data = pd.read_csv(zip_file.open(csv_file))\n"
     ]
    }
   ],
   "source": [
    "FILE = \"./italian_restaurants.zip\"\n",
    "zip_file = ZipFile(FILE)\n",
    "csv_file = zip_file.namelist()[0]\n",
    "data = pd.read_csv(zip_file.open(csv_file))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Unnamed: 0                                                                      650043\n",
       "restaurant_link                                                    g10029467-d18338571\n",
       "restaurant_name                                                                L'Ortal\n",
       "original_location                    [\"Europe\", \"Italy\", \"Friuli Venezia Giulia\", \"...\n",
       "country                                                                          Italy\n",
       "region                                                           Friuli Venezia Giulia\n",
       "province                                                         Province of Pordenone\n",
       "city                                                                               NaN\n",
       "address                                  Via Jacopo Ortis 15, 33090, Vito d'Asio Italy\n",
       "latitude                                                                      46.22801\n",
       "longitude                                                                      12.9397\n",
       "claimed                                                                      Unclaimed\n",
       "awards                                                                             NaN\n",
       "popularity_detailed                                  #1 of 1 Restaurant in Vito d'Asio\n",
       "popularity_generic                                #2 of 3 places to eat in Vito d'Asio\n",
       "top_tags                                         Cheap Eats, Italian, Northern-Italian\n",
       "price_level                                                                          â‚¬\n",
       "price_range                                                                        NaN\n",
       "meals                                                                              NaN\n",
       "cuisines                                                     Italian, Northern-Italian\n",
       "special_diets                                                                      NaN\n",
       "features                                                                           NaN\n",
       "vegetarian_friendly                                                                  N\n",
       "vegan_options                                                                        N\n",
       "gluten_free                                                                          N\n",
       "original_open_hours                                                                NaN\n",
       "open_days_per_week                                                                 NaN\n",
       "open_hours_per_week                                                                NaN\n",
       "working_shifts_per_week                                                            NaN\n",
       "avg_rating                                                                         5.0\n",
       "total_reviews_count                                                                5.0\n",
       "default_language                                                         All languages\n",
       "reviews_count_in_default_language                                                  5.0\n",
       "excellent                                                                          4.0\n",
       "very_good                                                                          1.0\n",
       "average                                                                            0.0\n",
       "poor                                                                               0.0\n",
       "terrible                                                                           0.0\n",
       "food                                                                               NaN\n",
       "service                                                                            NaN\n",
       "value                                                                              NaN\n",
       "atmosphere                                                                         NaN\n",
       "keywords                                                                           NaN\n",
       "Name: 0, dtype: object"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.iloc[0]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Removing useless fields"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = data.drop([\"claimed\", \n",
    "           \"awards\", \n",
    "           \"popularity_detailed\", \n",
    "           \"popularity_generic\", \n",
    "           \"price_range\", \n",
    "           \"features\", \n",
    "           \"open_days_per_week\", \n",
    "           \"open_hours_per_week\", \n",
    "           \"working_shifts_per_week\",\n",
    "           \"default_language\",\n",
    "           \"reviews_count_in_default_language\",\n",
    "           \"excellent\",\n",
    "           \"very_good\",\n",
    "           \"average\",\n",
    "           \"poor\",\n",
    "           \"terrible\",\n",
    "           \"food\",\n",
    "           \"service\",\n",
    "           \"value\",\n",
    "           \"atmosphere\",\n",
    "           \"keywords\"], axis=1)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Populating the \"City\" field taking it from the address"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "comuni_df = pd.read_csv(\"./elenco_comuni_italiani.csv\", encoding='iso-8859-1', delimiter=\";\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "257bb6f124834a24a78d68b8213a8b27",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/224763 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#TODO: \"Rome\" non viene preso, quindi risolvi questa cosa\n",
    "def build_automaton(keywords):\n",
    "    automaton = Automaton()\n",
    "    for keyword in keywords:\n",
    "        automaton.add_word(keyword, keyword)\n",
    "    automaton.make_automaton()\n",
    "    return automaton\n",
    "\n",
    "keywords = [row['Denominazione in italiano'].lower() for _, row in comuni_df.iterrows()]\n",
    "automaton = build_automaton(keywords)\n",
    "\n",
    "def find_comune(address):\n",
    "    found = set()\n",
    "    for item in automaton.iter(\",\".join(address.split(\",\")[1:]).lower()):\n",
    "        found.add(item[1])\n",
    "    if found:\n",
    "        return max(found, key=len)\n",
    "    else:\n",
    "        return None\n",
    "\n",
    "tqdm.pandas()\n",
    "df['city'] = df['address'].progress_apply(find_comune)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "final_df = df.copy()\n",
    "final_df['total_reviews_count'] = final_df['total_reviews_count'].fillna(0.0)\n",
    "final_df['avg_rating'] = final_df['avg_rating'].fillna(0.0)\n",
    "final_df = final_df.fillna(\"\")\n",
    "final_df = final_df[final_df[\"latitude\"] != \"\"]\n",
    "final_df = final_df[final_df[\"longitude\"] != \"\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>restaurant_link</th>\n",
       "      <th>restaurant_name</th>\n",
       "      <th>original_location</th>\n",
       "      <th>country</th>\n",
       "      <th>region</th>\n",
       "      <th>province</th>\n",
       "      <th>city</th>\n",
       "      <th>address</th>\n",
       "      <th>...</th>\n",
       "      <th>price_level</th>\n",
       "      <th>meals</th>\n",
       "      <th>cuisines</th>\n",
       "      <th>special_diets</th>\n",
       "      <th>vegetarian_friendly</th>\n",
       "      <th>vegan_options</th>\n",
       "      <th>gluten_free</th>\n",
       "      <th>original_open_hours</th>\n",
       "      <th>avg_rating</th>\n",
       "      <th>total_reviews_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>650043</td>\n",
       "      <td>g10029467-d18338571</td>\n",
       "      <td>L'Ortal</td>\n",
       "      <td>[\"Europe\", \"Italy\", \"Friuli Venezia Giulia\", \"...</td>\n",
       "      <td>Italy</td>\n",
       "      <td>Friuli Venezia Giulia</td>\n",
       "      <td>Province of Pordenone</td>\n",
       "      <td>vito d'asio</td>\n",
       "      <td>Via Jacopo Ortis 15, 33090, Vito d'Asio Italy</td>\n",
       "      <td>...</td>\n",
       "      <td>â‚¬</td>\n",
       "      <td></td>\n",
       "      <td>Italian, Northern-Italian</td>\n",
       "      <td></td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td></td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>650044</td>\n",
       "      <td>g10029467-d23200955</td>\n",
       "      <td>Osteria con Cucina Lorenzini</td>\n",
       "      <td>[\"Europe\", \"Italy\", \"Friuli Venezia Giulia\", \"...</td>\n",
       "      <td>Italy</td>\n",
       "      <td>Friuli Venezia Giulia</td>\n",
       "      <td>Province of Pordenone</td>\n",
       "      <td>vito d'asio</td>\n",
       "      <td>Via Chiamp, 18, 33090, Vito d'Asio Italy</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>Breakfast, Lunch, Dinner</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>{\"Mon\": [\"08:00-19:00\"], \"Tue\": [\"08:00-19:00\"...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>650045</td>\n",
       "      <td>g10029597-d6878487</td>\n",
       "      <td>Ristorante Da Renzo</td>\n",
       "      <td>[\"Europe\", \"Italy\", \"Friuli Venezia Giulia\", \"...</td>\n",
       "      <td>Italy</td>\n",
       "      <td>Friuli Venezia Giulia</td>\n",
       "      <td>Province of Pordenone</td>\n",
       "      <td>vito d'asio</td>\n",
       "      <td>Via Galantz 2, 33090 San Francesco, Vito d'Asi...</td>\n",
       "      <td>...</td>\n",
       "      <td>â‚¬</td>\n",
       "      <td>Lunch, Dinner, Breakfast</td>\n",
       "      <td>Italian</td>\n",
       "      <td>Vegetarian Friendly</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>{\"Mon\": [\"07:00-22:00\"], \"Tue\": [\"07:00-22:00\"...</td>\n",
       "      <td>4.5</td>\n",
       "      <td>65.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>650046</td>\n",
       "      <td>g10038815-d14177798</td>\n",
       "      <td>Pizza Point</td>\n",
       "      <td>[\"Europe\", \"Italy\", \"Trentino-Alto Adige\", \"Pr...</td>\n",
       "      <td>Italy</td>\n",
       "      <td>Trentino-Alto Adige</td>\n",
       "      <td>Province of South Tyrol</td>\n",
       "      <td>valdaora</td>\n",
       "      <td>Via del Borgo 4B, 39030 Valdaora di Sopra, Val...</td>\n",
       "      <td>...</td>\n",
       "      <td>â‚¬</td>\n",
       "      <td></td>\n",
       "      <td>Pizza</td>\n",
       "      <td></td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td></td>\n",
       "      <td>5.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>650047</td>\n",
       "      <td>g10045458-d12940567</td>\n",
       "      <td>Pizzeria Snoopy 2</td>\n",
       "      <td>[\"Europe\", \"Italy\", \"Emilia-Romagna\", \"Provinc...</td>\n",
       "      <td>Italy</td>\n",
       "      <td>Emilia-Romagna</td>\n",
       "      <td>Province of Bologna</td>\n",
       "      <td>dozza</td>\n",
       "      <td>Piazza Liberta' 14, 40060 Toscanella, Dozza Italy</td>\n",
       "      <td>...</td>\n",
       "      <td>â‚¬</td>\n",
       "      <td></td>\n",
       "      <td>Italian, Pizza</td>\n",
       "      <td></td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td></td>\n",
       "      <td>3.0</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>222598</th>\n",
       "      <td>224758</td>\n",
       "      <td>874801</td>\n",
       "      <td>g9885748-d12879403</td>\n",
       "      <td>La Locanda dell'Eremita</td>\n",
       "      <td>[\"Europe\", \"Italy\", \"Campania\", \"Province of C...</td>\n",
       "      <td>Italy</td>\n",
       "      <td>Campania</td>\n",
       "      <td>Province of Caserta</td>\n",
       "      <td>caserta</td>\n",
       "      <td>Via San Vitaliano 30, 81100 Casola, Caserta Italy</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Italian</td>\n",
       "      <td></td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td></td>\n",
       "      <td>4.0</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>222599</th>\n",
       "      <td>224759</td>\n",
       "      <td>874802</td>\n",
       "      <td>g9885748-d4002177</td>\n",
       "      <td>Ristorante Pizzeria La Fenice</td>\n",
       "      <td>[\"Europe\", \"Italy\", \"Campania\", \"Province of C...</td>\n",
       "      <td>Italy</td>\n",
       "      <td>Campania</td>\n",
       "      <td>Province of Caserta</td>\n",
       "      <td>caserta</td>\n",
       "      <td>Via Dottore Ferdinando Rossi 41, 81100 Casola,...</td>\n",
       "      <td>...</td>\n",
       "      <td>â‚¬</td>\n",
       "      <td>Lunch, Dinner, After-hours</td>\n",
       "      <td>Italian, Pizza, Mediterranean, Neapolitan, Cam...</td>\n",
       "      <td></td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>{\"Mon\": [\"17:30-23:45\"], \"Tue\": [\"18:30-23:00\"...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>190.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>222600</th>\n",
       "      <td>224760</td>\n",
       "      <td>874803</td>\n",
       "      <td>g9885748-d6905025</td>\n",
       "      <td>Agriturismo l'Eremo</td>\n",
       "      <td>[\"Europe\", \"Italy\", \"Campania\", \"Province of C...</td>\n",
       "      <td>Italy</td>\n",
       "      <td>Campania</td>\n",
       "      <td>Province of Caserta</td>\n",
       "      <td>caserta</td>\n",
       "      <td>Via San Vitaliano 34, 81100 Casola, Caserta Italy</td>\n",
       "      <td>...</td>\n",
       "      <td>â‚¬â‚¬-â‚¬â‚¬â‚¬</td>\n",
       "      <td>Lunch, Dinner</td>\n",
       "      <td>Italian, Mediterranean</td>\n",
       "      <td></td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>{\"Mon\": [\"11:00-00:00\"], \"Tue\": [\"11:00-00:00\"...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>117.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>222601</th>\n",
       "      <td>224761</td>\n",
       "      <td>874804</td>\n",
       "      <td>g9885970-d15265965</td>\n",
       "      <td>L'Etoile</td>\n",
       "      <td>[\"Europe\", \"Italy\", \"Friuli Venezia Giulia\", \"...</td>\n",
       "      <td>Italy</td>\n",
       "      <td>Friuli Venezia Giulia</td>\n",
       "      <td>Province of Pordenone</td>\n",
       "      <td>aviano</td>\n",
       "      <td>Viale San Giorgio 9, 33081 Castello d'Aviano, ...</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td></td>\n",
       "      <td>4.5</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>222602</th>\n",
       "      <td>224762</td>\n",
       "      <td>874805</td>\n",
       "      <td>g9885970-d23194598</td>\n",
       "      <td>Osteria Borta</td>\n",
       "      <td>[\"Europe\", \"Italy\", \"Friuli Venezia Giulia\", \"...</td>\n",
       "      <td>Italy</td>\n",
       "      <td>Friuli Venezia Giulia</td>\n",
       "      <td>Province of Pordenone</td>\n",
       "      <td>aviano</td>\n",
       "      <td>Piazzale Armando Diaz, 3, Castello d'Aviano, A...</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Bar, Cafe, Pub, Wine Bar</td>\n",
       "      <td></td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td></td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>222603 rows Ã— 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         index  Unnamed: 0      restaurant_link  \\\n",
       "0            0      650043  g10029467-d18338571   \n",
       "1            1      650044  g10029467-d23200955   \n",
       "2            2      650045   g10029597-d6878487   \n",
       "3            3      650046  g10038815-d14177798   \n",
       "4            4      650047  g10045458-d12940567   \n",
       "...        ...         ...                  ...   \n",
       "222598  224758      874801   g9885748-d12879403   \n",
       "222599  224759      874802    g9885748-d4002177   \n",
       "222600  224760      874803    g9885748-d6905025   \n",
       "222601  224761      874804   g9885970-d15265965   \n",
       "222602  224762      874805   g9885970-d23194598   \n",
       "\n",
       "                      restaurant_name  \\\n",
       "0                             L'Ortal   \n",
       "1        Osteria con Cucina Lorenzini   \n",
       "2                 Ristorante Da Renzo   \n",
       "3                         Pizza Point   \n",
       "4                   Pizzeria Snoopy 2   \n",
       "...                               ...   \n",
       "222598        La Locanda dell'Eremita   \n",
       "222599  Ristorante Pizzeria La Fenice   \n",
       "222600            Agriturismo l'Eremo   \n",
       "222601                       L'Etoile   \n",
       "222602                  Osteria Borta   \n",
       "\n",
       "                                        original_location country  \\\n",
       "0       [\"Europe\", \"Italy\", \"Friuli Venezia Giulia\", \"...   Italy   \n",
       "1       [\"Europe\", \"Italy\", \"Friuli Venezia Giulia\", \"...   Italy   \n",
       "2       [\"Europe\", \"Italy\", \"Friuli Venezia Giulia\", \"...   Italy   \n",
       "3       [\"Europe\", \"Italy\", \"Trentino-Alto Adige\", \"Pr...   Italy   \n",
       "4       [\"Europe\", \"Italy\", \"Emilia-Romagna\", \"Provinc...   Italy   \n",
       "...                                                   ...     ...   \n",
       "222598  [\"Europe\", \"Italy\", \"Campania\", \"Province of C...   Italy   \n",
       "222599  [\"Europe\", \"Italy\", \"Campania\", \"Province of C...   Italy   \n",
       "222600  [\"Europe\", \"Italy\", \"Campania\", \"Province of C...   Italy   \n",
       "222601  [\"Europe\", \"Italy\", \"Friuli Venezia Giulia\", \"...   Italy   \n",
       "222602  [\"Europe\", \"Italy\", \"Friuli Venezia Giulia\", \"...   Italy   \n",
       "\n",
       "                       region                 province         city  \\\n",
       "0       Friuli Venezia Giulia    Province of Pordenone  vito d'asio   \n",
       "1       Friuli Venezia Giulia    Province of Pordenone  vito d'asio   \n",
       "2       Friuli Venezia Giulia    Province of Pordenone  vito d'asio   \n",
       "3         Trentino-Alto Adige  Province of South Tyrol     valdaora   \n",
       "4              Emilia-Romagna      Province of Bologna        dozza   \n",
       "...                       ...                      ...          ...   \n",
       "222598               Campania      Province of Caserta      caserta   \n",
       "222599               Campania      Province of Caserta      caserta   \n",
       "222600               Campania      Province of Caserta      caserta   \n",
       "222601  Friuli Venezia Giulia    Province of Pordenone       aviano   \n",
       "222602  Friuli Venezia Giulia    Province of Pordenone       aviano   \n",
       "\n",
       "                                                  address  ... price_level  \\\n",
       "0           Via Jacopo Ortis 15, 33090, Vito d'Asio Italy  ...           â‚¬   \n",
       "1                Via Chiamp, 18, 33090, Vito d'Asio Italy  ...               \n",
       "2       Via Galantz 2, 33090 San Francesco, Vito d'Asi...  ...           â‚¬   \n",
       "3       Via del Borgo 4B, 39030 Valdaora di Sopra, Val...  ...           â‚¬   \n",
       "4       Piazza Liberta' 14, 40060 Toscanella, Dozza Italy  ...           â‚¬   \n",
       "...                                                   ...  ...         ...   \n",
       "222598  Via San Vitaliano 30, 81100 Casola, Caserta Italy  ...               \n",
       "222599  Via Dottore Ferdinando Rossi 41, 81100 Casola,...  ...           â‚¬   \n",
       "222600  Via San Vitaliano 34, 81100 Casola, Caserta Italy  ...      â‚¬â‚¬-â‚¬â‚¬â‚¬   \n",
       "222601  Viale San Giorgio 9, 33081 Castello d'Aviano, ...  ...               \n",
       "222602  Piazzale Armando Diaz, 3, Castello d'Aviano, A...  ...               \n",
       "\n",
       "                             meals  \\\n",
       "0                                    \n",
       "1         Breakfast, Lunch, Dinner   \n",
       "2         Lunch, Dinner, Breakfast   \n",
       "3                                    \n",
       "4                                    \n",
       "...                            ...   \n",
       "222598                               \n",
       "222599  Lunch, Dinner, After-hours   \n",
       "222600               Lunch, Dinner   \n",
       "222601                               \n",
       "222602                               \n",
       "\n",
       "                                                 cuisines  \\\n",
       "0                               Italian, Northern-Italian   \n",
       "1                                                           \n",
       "2                                                 Italian   \n",
       "3                                                   Pizza   \n",
       "4                                          Italian, Pizza   \n",
       "...                                                   ...   \n",
       "222598                                            Italian   \n",
       "222599  Italian, Pizza, Mediterranean, Neapolitan, Cam...   \n",
       "222600                             Italian, Mediterranean   \n",
       "222601                                                      \n",
       "222602                           Bar, Cafe, Pub, Wine Bar   \n",
       "\n",
       "              special_diets vegetarian_friendly vegan_options gluten_free  \\\n",
       "0                                             N             N           N   \n",
       "1                                             N             N           N   \n",
       "2       Vegetarian Friendly                   Y             N           N   \n",
       "3                                             N             N           N   \n",
       "4                                             N             N           N   \n",
       "...                     ...                 ...           ...         ...   \n",
       "222598                                        N             N           N   \n",
       "222599                                        N             N           N   \n",
       "222600                                        N             N           N   \n",
       "222601                                        N             N           N   \n",
       "222602                                        N             N           N   \n",
       "\n",
       "                                      original_open_hours avg_rating  \\\n",
       "0                                                                5.0   \n",
       "1       {\"Mon\": [\"08:00-19:00\"], \"Tue\": [\"08:00-19:00\"...        0.0   \n",
       "2       {\"Mon\": [\"07:00-22:00\"], \"Tue\": [\"07:00-22:00\"...        4.5   \n",
       "3                                                                5.0   \n",
       "4                                                                3.0   \n",
       "...                                                   ...        ...   \n",
       "222598                                                           4.0   \n",
       "222599  {\"Mon\": [\"17:30-23:45\"], \"Tue\": [\"18:30-23:00\"...        4.0   \n",
       "222600  {\"Mon\": [\"11:00-00:00\"], \"Tue\": [\"11:00-00:00\"...        4.0   \n",
       "222601                                                           4.5   \n",
       "222602                                                           0.0   \n",
       "\n",
       "       total_reviews_count  \n",
       "0                      5.0  \n",
       "1                      0.0  \n",
       "2                     65.0  \n",
       "3                      4.0  \n",
       "4                     13.0  \n",
       "...                    ...  \n",
       "222598                 7.0  \n",
       "222599               190.0  \n",
       "222600               117.0  \n",
       "222601                 3.0  \n",
       "222602                 0.0  \n",
       "\n",
       "[222603 rows x 23 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = final_df.reset_index()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Save the dataframe as a zipped file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "from zipfile import ZipFile\n",
    "import os\n",
    "\n",
    "DF_PATH = \"./italian_restaurant_clean.csv\"\n",
    "ZIP_PATH = \"./italian_restaurant_clean.zip\"\n",
    "final_df.to_csv(DF_PATH)\n",
    "with ZipFile(ZIP_PATH, \"w\") as myzip:\n",
    "    myzip.write(DF_PATH)\n",
    "\n",
    "os.remove(DF_PATH)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Code to insert the restaurants in the database "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\"./italian_restaurant_clean.zip\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "3ad1ef888d4c4301b097534d3cf0d96b",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/222603 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "ename": "ConnectionError",
     "evalue": "('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mRemoteDisconnected\u001b[0m                        Traceback (most recent call last)",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/urllib3/connectionpool.py:699\u001b[0m, in \u001b[0;36mHTTPConnectionPool.urlopen\u001b[0;34m(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, **response_kw)\u001b[0m\n\u001b[1;32m    698\u001b[0m \u001b[39m# Make the request on the httplib connection object.\u001b[39;00m\n\u001b[0;32m--> 699\u001b[0m httplib_response \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_make_request(\n\u001b[1;32m    700\u001b[0m     conn,\n\u001b[1;32m    701\u001b[0m     method,\n\u001b[1;32m    702\u001b[0m     url,\n\u001b[1;32m    703\u001b[0m     timeout\u001b[39m=\u001b[39;49mtimeout_obj,\n\u001b[1;32m    704\u001b[0m     body\u001b[39m=\u001b[39;49mbody,\n\u001b[1;32m    705\u001b[0m     headers\u001b[39m=\u001b[39;49mheaders,\n\u001b[1;32m    706\u001b[0m     chunked\u001b[39m=\u001b[39;49mchunked,\n\u001b[1;32m    707\u001b[0m )\n\u001b[1;32m    709\u001b[0m \u001b[39m# If we're going to release the connection in ``finally:``, then\u001b[39;00m\n\u001b[1;32m    710\u001b[0m \u001b[39m# the response doesn't need to know about the connection. Otherwise\u001b[39;00m\n\u001b[1;32m    711\u001b[0m \u001b[39m# it will also try to release it and we'll have a double-release\u001b[39;00m\n\u001b[1;32m    712\u001b[0m \u001b[39m# mess.\u001b[39;00m\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/urllib3/connectionpool.py:445\u001b[0m, in \u001b[0;36mHTTPConnectionPool._make_request\u001b[0;34m(self, conn, method, url, timeout, chunked, **httplib_request_kw)\u001b[0m\n\u001b[1;32m    441\u001b[0m         \u001b[39mexcept\u001b[39;00m \u001b[39mBaseException\u001b[39;00m \u001b[39mas\u001b[39;00m e:\n\u001b[1;32m    442\u001b[0m             \u001b[39m# Remove the TypeError from the exception chain in\u001b[39;00m\n\u001b[1;32m    443\u001b[0m             \u001b[39m# Python 3 (including for exceptions like SystemExit).\u001b[39;00m\n\u001b[1;32m    444\u001b[0m             \u001b[39m# Otherwise it looks like a bug in the code.\u001b[39;00m\n\u001b[0;32m--> 445\u001b[0m             six\u001b[39m.\u001b[39;49mraise_from(e, \u001b[39mNone\u001b[39;49;00m)\n\u001b[1;32m    446\u001b[0m \u001b[39mexcept\u001b[39;00m (SocketTimeout, BaseSSLError, SocketError) \u001b[39mas\u001b[39;00m e:\n",
      "File \u001b[0;32m<string>:3\u001b[0m, in \u001b[0;36mraise_from\u001b[0;34m(value, from_value)\u001b[0m\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/urllib3/connectionpool.py:440\u001b[0m, in \u001b[0;36mHTTPConnectionPool._make_request\u001b[0;34m(self, conn, method, url, timeout, chunked, **httplib_request_kw)\u001b[0m\n\u001b[1;32m    439\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 440\u001b[0m     httplib_response \u001b[39m=\u001b[39m conn\u001b[39m.\u001b[39;49mgetresponse()\n\u001b[1;32m    441\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mBaseException\u001b[39;00m \u001b[39mas\u001b[39;00m e:\n\u001b[1;32m    442\u001b[0m     \u001b[39m# Remove the TypeError from the exception chain in\u001b[39;00m\n\u001b[1;32m    443\u001b[0m     \u001b[39m# Python 3 (including for exceptions like SystemExit).\u001b[39;00m\n\u001b[1;32m    444\u001b[0m     \u001b[39m# Otherwise it looks like a bug in the code.\u001b[39;00m\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/http/client.py:1377\u001b[0m, in \u001b[0;36mHTTPConnection.getresponse\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1376\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m-> 1377\u001b[0m     response\u001b[39m.\u001b[39;49mbegin()\n\u001b[1;32m   1378\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mConnectionError\u001b[39;00m:\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/http/client.py:320\u001b[0m, in \u001b[0;36mHTTPResponse.begin\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    319\u001b[0m \u001b[39mwhile\u001b[39;00m \u001b[39mTrue\u001b[39;00m:\n\u001b[0;32m--> 320\u001b[0m     version, status, reason \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_read_status()\n\u001b[1;32m    321\u001b[0m     \u001b[39mif\u001b[39;00m status \u001b[39m!=\u001b[39m CONTINUE:\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/http/client.py:289\u001b[0m, in \u001b[0;36mHTTPResponse._read_status\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    286\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m line:\n\u001b[1;32m    287\u001b[0m     \u001b[39m# Presumably, the server closed the connection before\u001b[39;00m\n\u001b[1;32m    288\u001b[0m     \u001b[39m# sending a valid response.\u001b[39;00m\n\u001b[0;32m--> 289\u001b[0m     \u001b[39mraise\u001b[39;00m RemoteDisconnected(\u001b[39m\"\u001b[39m\u001b[39mRemote end closed connection without\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m    290\u001b[0m                              \u001b[39m\"\u001b[39m\u001b[39m response\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[1;32m    291\u001b[0m \u001b[39mtry\u001b[39;00m:\n",
      "\u001b[0;31mRemoteDisconnected\u001b[0m: Remote end closed connection without response",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mProtocolError\u001b[0m                             Traceback (most recent call last)",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/requests/adapters.py:440\u001b[0m, in \u001b[0;36mHTTPAdapter.send\u001b[0;34m(self, request, stream, timeout, verify, cert, proxies)\u001b[0m\n\u001b[1;32m    439\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m chunked:\n\u001b[0;32m--> 440\u001b[0m     resp \u001b[39m=\u001b[39m conn\u001b[39m.\u001b[39;49murlopen(\n\u001b[1;32m    441\u001b[0m         method\u001b[39m=\u001b[39;49mrequest\u001b[39m.\u001b[39;49mmethod,\n\u001b[1;32m    442\u001b[0m         url\u001b[39m=\u001b[39;49murl,\n\u001b[1;32m    443\u001b[0m         body\u001b[39m=\u001b[39;49mrequest\u001b[39m.\u001b[39;49mbody,\n\u001b[1;32m    444\u001b[0m         headers\u001b[39m=\u001b[39;49mrequest\u001b[39m.\u001b[39;49mheaders,\n\u001b[1;32m    445\u001b[0m         redirect\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m,\n\u001b[1;32m    446\u001b[0m         assert_same_host\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m,\n\u001b[1;32m    447\u001b[0m         preload_content\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m,\n\u001b[1;32m    448\u001b[0m         decode_content\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m,\n\u001b[1;32m    449\u001b[0m         retries\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mmax_retries,\n\u001b[1;32m    450\u001b[0m         timeout\u001b[39m=\u001b[39;49mtimeout\n\u001b[1;32m    451\u001b[0m     )\n\u001b[1;32m    453\u001b[0m \u001b[39m# Send the request.\u001b[39;00m\n\u001b[1;32m    454\u001b[0m \u001b[39melse\u001b[39;00m:\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/urllib3/connectionpool.py:755\u001b[0m, in \u001b[0;36mHTTPConnectionPool.urlopen\u001b[0;34m(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, **response_kw)\u001b[0m\n\u001b[1;32m    753\u001b[0m     e \u001b[39m=\u001b[39m ProtocolError(\u001b[39m\"\u001b[39m\u001b[39mConnection aborted.\u001b[39m\u001b[39m\"\u001b[39m, e)\n\u001b[0;32m--> 755\u001b[0m retries \u001b[39m=\u001b[39m retries\u001b[39m.\u001b[39;49mincrement(\n\u001b[1;32m    756\u001b[0m     method, url, error\u001b[39m=\u001b[39;49me, _pool\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m, _stacktrace\u001b[39m=\u001b[39;49msys\u001b[39m.\u001b[39;49mexc_info()[\u001b[39m2\u001b[39;49m]\n\u001b[1;32m    757\u001b[0m )\n\u001b[1;32m    758\u001b[0m retries\u001b[39m.\u001b[39msleep()\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/urllib3/util/retry.py:531\u001b[0m, in \u001b[0;36mRetry.increment\u001b[0;34m(self, method, url, response, error, _pool, _stacktrace)\u001b[0m\n\u001b[1;32m    530\u001b[0m \u001b[39mif\u001b[39;00m read \u001b[39mis\u001b[39;00m \u001b[39mFalse\u001b[39;00m \u001b[39mor\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_is_method_retryable(method):\n\u001b[0;32m--> 531\u001b[0m     \u001b[39mraise\u001b[39;00m six\u001b[39m.\u001b[39;49mreraise(\u001b[39mtype\u001b[39;49m(error), error, _stacktrace)\n\u001b[1;32m    532\u001b[0m \u001b[39melif\u001b[39;00m read \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/urllib3/packages/six.py:734\u001b[0m, in \u001b[0;36mreraise\u001b[0;34m(tp, value, tb)\u001b[0m\n\u001b[1;32m    733\u001b[0m \u001b[39mif\u001b[39;00m value\u001b[39m.\u001b[39m__traceback__ \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m tb:\n\u001b[0;32m--> 734\u001b[0m     \u001b[39mraise\u001b[39;00m value\u001b[39m.\u001b[39mwith_traceback(tb)\n\u001b[1;32m    735\u001b[0m \u001b[39mraise\u001b[39;00m value\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/urllib3/connectionpool.py:699\u001b[0m, in \u001b[0;36mHTTPConnectionPool.urlopen\u001b[0;34m(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, **response_kw)\u001b[0m\n\u001b[1;32m    698\u001b[0m \u001b[39m# Make the request on the httplib connection object.\u001b[39;00m\n\u001b[0;32m--> 699\u001b[0m httplib_response \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_make_request(\n\u001b[1;32m    700\u001b[0m     conn,\n\u001b[1;32m    701\u001b[0m     method,\n\u001b[1;32m    702\u001b[0m     url,\n\u001b[1;32m    703\u001b[0m     timeout\u001b[39m=\u001b[39;49mtimeout_obj,\n\u001b[1;32m    704\u001b[0m     body\u001b[39m=\u001b[39;49mbody,\n\u001b[1;32m    705\u001b[0m     headers\u001b[39m=\u001b[39;49mheaders,\n\u001b[1;32m    706\u001b[0m     chunked\u001b[39m=\u001b[39;49mchunked,\n\u001b[1;32m    707\u001b[0m )\n\u001b[1;32m    709\u001b[0m \u001b[39m# If we're going to release the connection in ``finally:``, then\u001b[39;00m\n\u001b[1;32m    710\u001b[0m \u001b[39m# the response doesn't need to know about the connection. Otherwise\u001b[39;00m\n\u001b[1;32m    711\u001b[0m \u001b[39m# it will also try to release it and we'll have a double-release\u001b[39;00m\n\u001b[1;32m    712\u001b[0m \u001b[39m# mess.\u001b[39;00m\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/urllib3/connectionpool.py:445\u001b[0m, in \u001b[0;36mHTTPConnectionPool._make_request\u001b[0;34m(self, conn, method, url, timeout, chunked, **httplib_request_kw)\u001b[0m\n\u001b[1;32m    441\u001b[0m         \u001b[39mexcept\u001b[39;00m \u001b[39mBaseException\u001b[39;00m \u001b[39mas\u001b[39;00m e:\n\u001b[1;32m    442\u001b[0m             \u001b[39m# Remove the TypeError from the exception chain in\u001b[39;00m\n\u001b[1;32m    443\u001b[0m             \u001b[39m# Python 3 (including for exceptions like SystemExit).\u001b[39;00m\n\u001b[1;32m    444\u001b[0m             \u001b[39m# Otherwise it looks like a bug in the code.\u001b[39;00m\n\u001b[0;32m--> 445\u001b[0m             six\u001b[39m.\u001b[39;49mraise_from(e, \u001b[39mNone\u001b[39;49;00m)\n\u001b[1;32m    446\u001b[0m \u001b[39mexcept\u001b[39;00m (SocketTimeout, BaseSSLError, SocketError) \u001b[39mas\u001b[39;00m e:\n",
      "File \u001b[0;32m<string>:3\u001b[0m, in \u001b[0;36mraise_from\u001b[0;34m(value, from_value)\u001b[0m\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/urllib3/connectionpool.py:440\u001b[0m, in \u001b[0;36mHTTPConnectionPool._make_request\u001b[0;34m(self, conn, method, url, timeout, chunked, **httplib_request_kw)\u001b[0m\n\u001b[1;32m    439\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 440\u001b[0m     httplib_response \u001b[39m=\u001b[39m conn\u001b[39m.\u001b[39;49mgetresponse()\n\u001b[1;32m    441\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mBaseException\u001b[39;00m \u001b[39mas\u001b[39;00m e:\n\u001b[1;32m    442\u001b[0m     \u001b[39m# Remove the TypeError from the exception chain in\u001b[39;00m\n\u001b[1;32m    443\u001b[0m     \u001b[39m# Python 3 (including for exceptions like SystemExit).\u001b[39;00m\n\u001b[1;32m    444\u001b[0m     \u001b[39m# Otherwise it looks like a bug in the code.\u001b[39;00m\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/http/client.py:1377\u001b[0m, in \u001b[0;36mHTTPConnection.getresponse\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1376\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m-> 1377\u001b[0m     response\u001b[39m.\u001b[39;49mbegin()\n\u001b[1;32m   1378\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mConnectionError\u001b[39;00m:\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/http/client.py:320\u001b[0m, in \u001b[0;36mHTTPResponse.begin\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    319\u001b[0m \u001b[39mwhile\u001b[39;00m \u001b[39mTrue\u001b[39;00m:\n\u001b[0;32m--> 320\u001b[0m     version, status, reason \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_read_status()\n\u001b[1;32m    321\u001b[0m     \u001b[39mif\u001b[39;00m status \u001b[39m!=\u001b[39m CONTINUE:\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/http/client.py:289\u001b[0m, in \u001b[0;36mHTTPResponse._read_status\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    286\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m line:\n\u001b[1;32m    287\u001b[0m     \u001b[39m# Presumably, the server closed the connection before\u001b[39;00m\n\u001b[1;32m    288\u001b[0m     \u001b[39m# sending a valid response.\u001b[39;00m\n\u001b[0;32m--> 289\u001b[0m     \u001b[39mraise\u001b[39;00m RemoteDisconnected(\u001b[39m\"\u001b[39m\u001b[39mRemote end closed connection without\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m    290\u001b[0m                              \u001b[39m\"\u001b[39m\u001b[39m response\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[1;32m    291\u001b[0m \u001b[39mtry\u001b[39;00m:\n",
      "\u001b[0;31mProtocolError\u001b[0m: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mConnectionError\u001b[0m                           Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[16], line 34\u001b[0m\n\u001b[1;32m     31\u001b[0m     result \u001b[39m=\u001b[39m requests\u001b[39m.\u001b[39mpost(URL, json\u001b[39m=\u001b[39mbody)\n\u001b[1;32m     33\u001b[0m tqdm\u001b[39m.\u001b[39mpandas()\n\u001b[0;32m---> 34\u001b[0m df\u001b[39m.\u001b[39;49mprogress_apply(create_restaurant, axis\u001b[39m=\u001b[39;49m\u001b[39m1\u001b[39;49m)\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/tqdm/std.py:814\u001b[0m, in \u001b[0;36mtqdm.pandas.<locals>.inner_generator.<locals>.inner\u001b[0;34m(df, func, *args, **kwargs)\u001b[0m\n\u001b[1;32m    811\u001b[0m \u001b[39m# Apply the provided function (in **kwargs)\u001b[39;00m\n\u001b[1;32m    812\u001b[0m \u001b[39m# on the df using our wrapper (which provides bar updating)\u001b[39;00m\n\u001b[1;32m    813\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 814\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mgetattr\u001b[39;49m(df, df_function)(wrapper, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n\u001b[1;32m    815\u001b[0m \u001b[39mfinally\u001b[39;00m:\n\u001b[1;32m    816\u001b[0m     t\u001b[39m.\u001b[39mclose()\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/pandas/core/frame.py:9565\u001b[0m, in \u001b[0;36mDataFrame.apply\u001b[0;34m(self, func, axis, raw, result_type, args, **kwargs)\u001b[0m\n\u001b[1;32m   9554\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mpandas\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mcore\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mapply\u001b[39;00m \u001b[39mimport\u001b[39;00m frame_apply\n\u001b[1;32m   9556\u001b[0m op \u001b[39m=\u001b[39m frame_apply(\n\u001b[1;32m   9557\u001b[0m     \u001b[39mself\u001b[39m,\n\u001b[1;32m   9558\u001b[0m     func\u001b[39m=\u001b[39mfunc,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   9563\u001b[0m     kwargs\u001b[39m=\u001b[39mkwargs,\n\u001b[1;32m   9564\u001b[0m )\n\u001b[0;32m-> 9565\u001b[0m \u001b[39mreturn\u001b[39;00m op\u001b[39m.\u001b[39;49mapply()\u001b[39m.\u001b[39m__finalize__(\u001b[39mself\u001b[39m, method\u001b[39m=\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mapply\u001b[39m\u001b[39m\"\u001b[39m)\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/pandas/core/apply.py:746\u001b[0m, in \u001b[0;36mFrameApply.apply\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    743\u001b[0m \u001b[39melif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mraw:\n\u001b[1;32m    744\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mapply_raw()\n\u001b[0;32m--> 746\u001b[0m \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mapply_standard()\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/pandas/core/apply.py:873\u001b[0m, in \u001b[0;36mFrameApply.apply_standard\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    872\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mapply_standard\u001b[39m(\u001b[39mself\u001b[39m):\n\u001b[0;32m--> 873\u001b[0m     results, res_index \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mapply_series_generator()\n\u001b[1;32m    875\u001b[0m     \u001b[39m# wrap results\u001b[39;00m\n\u001b[1;32m    876\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mwrap_results(results, res_index)\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/pandas/core/apply.py:889\u001b[0m, in \u001b[0;36mFrameApply.apply_series_generator\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    886\u001b[0m \u001b[39mwith\u001b[39;00m option_context(\u001b[39m\"\u001b[39m\u001b[39mmode.chained_assignment\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39mNone\u001b[39;00m):\n\u001b[1;32m    887\u001b[0m     \u001b[39mfor\u001b[39;00m i, v \u001b[39min\u001b[39;00m \u001b[39menumerate\u001b[39m(series_gen):\n\u001b[1;32m    888\u001b[0m         \u001b[39m# ignore SettingWithCopy here in case the user mutates\u001b[39;00m\n\u001b[0;32m--> 889\u001b[0m         results[i] \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mf(v)\n\u001b[1;32m    890\u001b[0m         \u001b[39mif\u001b[39;00m \u001b[39misinstance\u001b[39m(results[i], ABCSeries):\n\u001b[1;32m    891\u001b[0m             \u001b[39m# If we have a view on v, we need to make a copy because\u001b[39;00m\n\u001b[1;32m    892\u001b[0m             \u001b[39m#  series_generator will swap out the underlying data\u001b[39;00m\n\u001b[1;32m    893\u001b[0m             results[i] \u001b[39m=\u001b[39m results[i]\u001b[39m.\u001b[39mcopy(deep\u001b[39m=\u001b[39m\u001b[39mFalse\u001b[39;00m)\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/tqdm/std.py:809\u001b[0m, in \u001b[0;36mtqdm.pandas.<locals>.inner_generator.<locals>.inner.<locals>.wrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    803\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mwrapper\u001b[39m(\u001b[39m*\u001b[39margs, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs):\n\u001b[1;32m    804\u001b[0m     \u001b[39m# update tbar correctly\u001b[39;00m\n\u001b[1;32m    805\u001b[0m     \u001b[39m# it seems `pandas apply` calls `func` twice\u001b[39;00m\n\u001b[1;32m    806\u001b[0m     \u001b[39m# on the first column/row to decide whether it can\u001b[39;00m\n\u001b[1;32m    807\u001b[0m     \u001b[39m# take a fast or slow code path; so stop when t.total==t.n\u001b[39;00m\n\u001b[1;32m    808\u001b[0m     t\u001b[39m.\u001b[39mupdate(n\u001b[39m=\u001b[39m\u001b[39m1\u001b[39m \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m t\u001b[39m.\u001b[39mtotal \u001b[39mor\u001b[39;00m t\u001b[39m.\u001b[39mn \u001b[39m<\u001b[39m t\u001b[39m.\u001b[39mtotal \u001b[39melse\u001b[39;00m \u001b[39m0\u001b[39m)\n\u001b[0;32m--> 809\u001b[0m     \u001b[39mreturn\u001b[39;00m func(\u001b[39m*\u001b[39;49margs, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n",
      "Cell \u001b[0;32mIn[16], line 31\u001b[0m, in \u001b[0;36mcreate_restaurant\u001b[0;34m(row)\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mcreate_restaurant\u001b[39m(row):\n\u001b[1;32m      8\u001b[0m     body \u001b[39m=\u001b[39m {\n\u001b[1;32m      9\u001b[0m             \u001b[39m\"\u001b[39m\u001b[39mname\u001b[39m\u001b[39m\"\u001b[39m: row[\u001b[39m\"\u001b[39m\u001b[39mrestaurant_name\u001b[39m\u001b[39m\"\u001b[39m],\n\u001b[1;32m     10\u001b[0m             \u001b[39m\"\u001b[39m\u001b[39mimageName\u001b[39m\u001b[39m\"\u001b[39m: \u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mrestaurant_image_\u001b[39m\u001b[39m{\u001b[39;00massign_image(row[\u001b[39m'\u001b[39m\u001b[39mindex\u001b[39m\u001b[39m'\u001b[39m])\u001b[39m}\u001b[39;00m\u001b[39m\"\u001b[39m,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     27\u001b[0m             \u001b[39m\"\u001b[39m\u001b[39mreviewsNumber\u001b[39m\u001b[39m\"\u001b[39m: \u001b[39mint\u001b[39m(row[\u001b[39m\"\u001b[39m\u001b[39mtotal_reviews_count\u001b[39m\u001b[39m\"\u001b[39m])\n\u001b[1;32m     28\u001b[0m         }\n\u001b[0;32m---> 31\u001b[0m     result \u001b[39m=\u001b[39m requests\u001b[39m.\u001b[39;49mpost(URL, json\u001b[39m=\u001b[39;49mbody)\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/requests/api.py:117\u001b[0m, in \u001b[0;36mpost\u001b[0;34m(url, data, json, **kwargs)\u001b[0m\n\u001b[1;32m    105\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mpost\u001b[39m(url, data\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m, json\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs):\n\u001b[1;32m    106\u001b[0m \u001b[39m    \u001b[39m\u001b[39mr\u001b[39m\u001b[39m\"\"\"Sends a POST request.\u001b[39;00m\n\u001b[1;32m    107\u001b[0m \n\u001b[1;32m    108\u001b[0m \u001b[39m    :param url: URL for the new :class:`Request` object.\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    114\u001b[0m \u001b[39m    :rtype: requests.Response\u001b[39;00m\n\u001b[1;32m    115\u001b[0m \u001b[39m    \"\"\"\u001b[39;00m\n\u001b[0;32m--> 117\u001b[0m     \u001b[39mreturn\u001b[39;00m request(\u001b[39m'\u001b[39;49m\u001b[39mpost\u001b[39;49m\u001b[39m'\u001b[39;49m, url, data\u001b[39m=\u001b[39;49mdata, json\u001b[39m=\u001b[39;49mjson, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/requests/api.py:61\u001b[0m, in \u001b[0;36mrequest\u001b[0;34m(method, url, **kwargs)\u001b[0m\n\u001b[1;32m     57\u001b[0m \u001b[39m# By using the 'with' statement we are sure the session is closed, thus we\u001b[39;00m\n\u001b[1;32m     58\u001b[0m \u001b[39m# avoid leaving sockets open which can trigger a ResourceWarning in some\u001b[39;00m\n\u001b[1;32m     59\u001b[0m \u001b[39m# cases, and look like a memory leak in others.\u001b[39;00m\n\u001b[1;32m     60\u001b[0m \u001b[39mwith\u001b[39;00m sessions\u001b[39m.\u001b[39mSession() \u001b[39mas\u001b[39;00m session:\n\u001b[0;32m---> 61\u001b[0m     \u001b[39mreturn\u001b[39;00m session\u001b[39m.\u001b[39;49mrequest(method\u001b[39m=\u001b[39;49mmethod, url\u001b[39m=\u001b[39;49murl, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/requests/sessions.py:529\u001b[0m, in \u001b[0;36mSession.request\u001b[0;34m(self, method, url, params, data, headers, cookies, files, auth, timeout, allow_redirects, proxies, hooks, stream, verify, cert, json)\u001b[0m\n\u001b[1;32m    524\u001b[0m send_kwargs \u001b[39m=\u001b[39m {\n\u001b[1;32m    525\u001b[0m     \u001b[39m'\u001b[39m\u001b[39mtimeout\u001b[39m\u001b[39m'\u001b[39m: timeout,\n\u001b[1;32m    526\u001b[0m     \u001b[39m'\u001b[39m\u001b[39mallow_redirects\u001b[39m\u001b[39m'\u001b[39m: allow_redirects,\n\u001b[1;32m    527\u001b[0m }\n\u001b[1;32m    528\u001b[0m send_kwargs\u001b[39m.\u001b[39mupdate(settings)\n\u001b[0;32m--> 529\u001b[0m resp \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49msend(prep, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49msend_kwargs)\n\u001b[1;32m    531\u001b[0m \u001b[39mreturn\u001b[39;00m resp\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/requests/sessions.py:645\u001b[0m, in \u001b[0;36mSession.send\u001b[0;34m(self, request, **kwargs)\u001b[0m\n\u001b[1;32m    642\u001b[0m start \u001b[39m=\u001b[39m preferred_clock()\n\u001b[1;32m    644\u001b[0m \u001b[39m# Send the request\u001b[39;00m\n\u001b[0;32m--> 645\u001b[0m r \u001b[39m=\u001b[39m adapter\u001b[39m.\u001b[39;49msend(request, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n\u001b[1;32m    647\u001b[0m \u001b[39m# Total elapsed time of the request (approximately)\u001b[39;00m\n\u001b[1;32m    648\u001b[0m elapsed \u001b[39m=\u001b[39m preferred_clock() \u001b[39m-\u001b[39m start\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/requests/adapters.py:501\u001b[0m, in \u001b[0;36mHTTPAdapter.send\u001b[0;34m(self, request, stream, timeout, verify, cert, proxies)\u001b[0m\n\u001b[1;32m    498\u001b[0m             \u001b[39mraise\u001b[39;00m\n\u001b[1;32m    500\u001b[0m \u001b[39mexcept\u001b[39;00m (ProtocolError, socket\u001b[39m.\u001b[39merror) \u001b[39mas\u001b[39;00m err:\n\u001b[0;32m--> 501\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mConnectionError\u001b[39;00m(err, request\u001b[39m=\u001b[39mrequest)\n\u001b[1;32m    503\u001b[0m \u001b[39mexcept\u001b[39;00m MaxRetryError \u001b[39mas\u001b[39;00m e:\n\u001b[1;32m    504\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39misinstance\u001b[39m(e\u001b[39m.\u001b[39mreason, ConnectTimeoutError):\n\u001b[1;32m    505\u001b[0m         \u001b[39m# TODO: Remove this in 3.0.0: see #2811\u001b[39;00m\n",
      "\u001b[0;31mConnectionError\u001b[0m: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))"
     ]
    }
   ],
   "source": [
    "URL = 'http://localhost:3000/restaurants/create-restaurant'\n",
    "\n",
    "def assign_image(index):\n",
    "    TOTAL_NUMBER_OF_IMAGES = 4761\n",
    "    return index % TOTAL_NUMBER_OF_IMAGES\n",
    "\n",
    "def create_restaurant(row):\n",
    "    body = {\n",
    "            \"name\": row[\"restaurant_name\"],\n",
    "            \"imageName\": f\"restaurant_image_{assign_image(row['index'])}\",\n",
    "            \"address\": row[\"address\"],\n",
    "            \"latitude\": float(row[\"latitude\"]),\n",
    "            \"longitude\": float(row[\"longitude\"]),\n",
    "            \"country\": row[\"country\"],\n",
    "            \"region\": row[\"region\"],\n",
    "            \"province\": row[\"province\"],\n",
    "            \"city\": row[\"city\"],\n",
    "            \"tags\": row[\"top_tags\"],\n",
    "            \"cuisines\": row[\"cuisines\"],\n",
    "            \"specialDiets\": row[\"special_diets\"],\n",
    "            \"priceLevel\": row[\"price_level\"],\n",
    "            \"meals\": row[\"meals\"],\n",
    "            \"avgRating\": float(row[\"avg_rating\"]),\n",
    "            \"vegetarianFriendly\": row[\"vegetarian_friendly\"] == \"Y\",\n",
    "            \"veganFriendly\": row[\"vegan_options\"] == \"Y\",\n",
    "            \"glutenFree\": row[\"gluten_free\"] == \"Y\",\n",
    "            \"reviewsNumber\": int(row[\"total_reviews_count\"])\n",
    "        }\n",
    "\n",
    "    \n",
    "    result = requests.post(URL, json=body)\n",
    "\n",
    "tqdm.pandas()\n",
    "df.progress_apply(create_restaurant, axis=1)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
